filestream/filestream.go

```diff
@@ -0,0 +1,106 @@
+package filestream
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/fsnotify/fsnotify"
+)
+
+type FileStream struct {
+	file_name string
+	reader    *os.File
+	watcher   *fsnotify.Watcher
+
+	// Used to abort streaming when a client disconnects, etc
+	ctx context.Context
+}
+
+//-------------------------------------------------------------------------------------------------
+func New(ctx context.Context, file_name string) (*FileStream, error) {
+	file, err := os.Open(file_name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open file '%s': %w", file_name, err)
+	}
+
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, fmt.Errorf("failed to start a watcher: %w", err)
+	}
+
+	err = watcher.Add(file_name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to add file '%s' to the watcher: %w", file_name, err)
+	}
+
+	return &FileStream{
+		file_name: file_name,
+		reader:    file,
+		watcher:   watcher,
+		ctx:       ctx,
+	}, nil
+}
+
+//-------------------------------------------------------------------------------------------------
+// Closes the fsnotify watcher and the underlying file stream
+func (s *FileStream) Close() error {
+	err := s.watcher.Close()
+	if err == nil {
+		err = s.reader.Close()
+	}
+	return err
+}
+
+//-------------------------------------------------------------------------------------------------
+func (s *FileStream) WaitForChanges() (changed bool) {
+	result := make(chan bool)
+	go func() {
+		for {
```

Comment:
What would happen if we are in this loop and `Close` is called but `s.ctx` is not cancelled?

Reply:
Good question. I was planning to later add functionality to allow Close() calls to stop the reader (since I'll need it to notify all readers when a command is done). Decided to add it now while addressing the double-close question as well.